scapy/contrib/roce.py

```diff
@@ -188,10 +207,35 @@ def compute_icrc(self, p):
     def post_build(self, p, pay):
         # type: (bytes, bytes) -> bytes
         p += pay
+        if self.opcode is None:
+            opcode = self.guess_opcode() or 0
+            p = struct.pack('B', opcode) + p[1:]
         if self.icrc is None:
             p = p[:-4] + self.compute_icrc(p)
         return p
 
+    def guess_opcode(self):
+        # type: () -> Optional[int]
+        'Guess the opcode based on the following layers.'
+        if isinstance(self.payload, DETH):
+            if isinstance(self.payload.payload, Immdt):
+                return opcode('UD', 'SEND_ONLY_WITH_IMMEDIATE')[0]
+            return opcode('UD', 'SEND_ONLY')[0]
+
+    def guess_payload_class(self, payload):
+        # type: (bytes) -> Type[Packet]
+        opcode = self.opcode
+        if transport(opcode) == _transports['RC'] or \
+           transport(opcode) == _transports['UC']:
+            cur_op = op(opcode)
```

Comment:
Since you are the original author, I take this opportunity to add that `opcode` really is superfluous. It could have easily been replaced with the raw ints, while only keeping the real names in comments for clarity. There is a lot of overhead here.

Reply:
I see. Do you mean overhead in terms of code or performance? Do you think it should be removed? I thought it would make it simpler for people to understand a dissected packet or build one.

Reply:
In terms of code.
Since you are using an `EnumField`, the op codes are already displayed nicely and  Scapy will also auto translate it if you input their string values while building a packet, which is in my opinion well enough.